src/NextjsDistribution.ts

```diff
@@ -461,15 +450,18 @@ export class NextjsDistribution extends Construct {
       domainNames,
       certificate: this.certificate,
       defaultBehavior: {
-        origin: fallbackOriginGroup, // try S3 first, then lambda
+        origin: new origins.HttpOrigin(Fn.parseDomainName(fnUrl.url), {
```

Comment:
@kevin-mitchell, why are you changing the default origin behavior? This seems to be linked to the S3 XML error message you're seeing that you described.

Reply:
@bestickley I changed it _to_ fix the issue with the 404/403s. Basically I was getting 403 errors that looked like S3 issues _before_ I changed this. Then I was trying to figure out what was going on, noted that what SST has for their defaultBehavior origin was quite different from this project, I updated to roughly match the simpler config they had and the issue went away.

THAT SAID, I am absolutely _not_ advocating for this change, I'm sure there is a great reason the origin was setup with the origin group like it was originally with the fallback origin group with the S3 fallback. I honestly just wasn't and am not sure what it is supposed to do. 

It seems the fallback default origin setup was (maybe?) for static resources, so that if the Next.js app returned a 404 that the S3 bucket would be checked for any static assets associated with it. Maybe that logic still makes sense, or maybe because of the new caching "system" in place it is adequate because the server already is checking the cache bucket for a cached resource.

Reply:
@kevin-mitchell, got it. Thanks for explaining. This change will likely break support for Lambda Function URL IAM Auth since the URL needs to be signed, so I'd prefer to keep it like it was. It seems to me like using an `HttpOrigin` of the function URL or making the primary origin the function origin should result in the same thing.

Do you know why the previous configuration wasn't working?

Reply:
@kevin-mitchell, I think that comment is wrong. See the original comment above the declaration of `fallbackOriginGroup`:

> // default handler for requests that don't match any other path:
    //   - try lambda handler first (/some-page, etc...)
    //   - if 403, fall back to S3
    //   - if 404, fall back to lambda handler
    //   - if 503, fall back to lambda handler

Since SST doesn't use fallback behavior to get assets from S3, I think we can assume that the server handler will handle fetching assets from S3 (but not 100% sure). So how about setting `defaultBehavior.origin` to `serverFunctionOrigin` to handle all traffic?

Reply:
@kevin-mitchell, I apologize. You're correct. I didn't see the definition of `serverFunctionOrigin` when reviewing. We should be all good to go.

Reply:
So after all of this, I reverted this change (see: https://github.com/jetbridge/cdk-nextjs/pull/125/commits/cec5e3232d8976746b7575c951f405a4804aa731)

I didn't realize before, but SST is generating the additional behaviors to cover the `/public` files at runtime. That's why they don't have the fallback. Rather than add confusion / noise, I'll leave the existing behavior for now untouched.

Reply:
Personally I do not like the SST approach to /public file handling because it creates a behavior for each file. It doesn't scale. I think it means if you have over 25 files in the root of /public your app cannot be deployed. Also it has bugs like if your file has a character like `(` it won't deploy either.

Reply:
It's definitely a trade off at build time vs runtime. You have to eat the 💩 🥪 either way.

Reply:
I also prefer this more elegant "if you're using AWS, use AWS" approach. I think. :).

That said I am still uncertain as to why the 404 is served as a 403 from CloudFront.

I _think_ I deployed a clean / fresh version of this project's `main` branch, along with `open-next` v 1.4, and I'm fairly certain I got the same 404 behavior. Any chance somebody would be willing or able to sanity check me? Is there something "special" about this test https://github.com/serverless-stack/open-next/tree/main/example (this is what I'm testing with currently).

Reply:
But remember, there's also a runtime penalty: 1) run nextjs server, page doesn't exist 2) look at S3 doesn't exist 3) fallback to 404. It might be negligible though.

Reply:
Question: Is there a world in which either:

1. I add a third fallback behavior, so the behavior is

Hit ServerHandler, if 403/404/503
Fallback to S3, if 403/404/503
Fall back to ServerHandler again

The original comment @bestickley called out makes it seem like this is how it is SUPPOSED to work already:

```
// default handler for requests that don't match any other path:
// - try lambda handler first (/some-page, etc...)
// - if 403, fall back to S3
// - if 404, fall back to lambda handler
// - if 503, fall back to lambda handler
```
but at least for me it seems like the fallback of S3 just fails and that's the end of it

_or_

2. Make the S3 bucket the first origin, then fallback to ServerHandler?

The second option is a bit expensive because it adds overhead of checking S3 to every single request.

Reply:
I designed CDK-NextJS to first hit S3, then if nothing exists, fall back to lambda.
However, this fallback handler is only used for routes that can't quickly be identified and routed to the correct place. For example if the route is under /api, /_next/image, etc then we don't need the fallback behavior.  So API and static files should be unaffected. If you have more information about your app, you can add your own `additionalBehaviors` as well to bypass the fallback origin system. You can also tell cloudfront to cache responses (perhaps with the new stale-while-revalidate) so it doesn't have to go to the origin every time.

I haven't measured the latency but we're talking about shaving off, I would guess, very low double-digit milliseconds in only a subset of requests.

Reply:
iirc, we did do the S3 first but reversed it due to the latency.

Reply:
> iirc, we did do the S3 first but reversed it due to the latency.

Yeah I thought I remembered doing S3 first, then went to check the code, and saw that is not presently the case...

Does the next handler return 403? Also doesn't the latest version of SST use a 502 or something instead? Any idea why?

Reply:
I honestly don't remember what caused the problem, but I recall getting about ~300ms in extra latency because of the fallback stuff.

open-next return 503 to fail immediately I think: https://github.com/serverless-stack/open-next/blob/main/packages/open-next/src/adapters/server-adapter.ts#L317

Per Frank: https://discord.com/channels/983865673656705025/1027265626085019769/1095464697978957875

Reply:
https://github.com/jetbridge/cdk-nextjs/blame/d3cee987d24674b0ccbc43a713ad665ddef22b15/src/NextjsDistribution.ts#L336-L344

vs

Reply:
Can you try w/ S3 first then server to see if you can reproduce the latency issue?

Reply:
I can look. I suspect it's a real thing. Even if it's a tiny bit of overhead, it's on every request. I also want to try out the app-sandbox to make sure I see the same behavior.

For larger / production / "enterprise" situations I would think people tune this behavior to optimize for what's most important to them.

Reply:
lol, very impressive chatGPT scifi! Off tangent, if you like hard sci fi and would like to alpha read my novel, hit me up!

Reply:
> I can look. I suspect it's a real thing. Even if it's a tiny bit of overhead, it's on every request. I also want to try out the app-sandbox to make sure I see the same behavior.

It's not every request; the static files and API paths don't have fallback origins.

Another option is to create an edge function that does something smart, like looking at the manifests NextJS spits out to determine where to route the request. Should be possible and likely similar to what Vercel does.

Reply:
> It's not every request; the static files and API paths don't have fallback origins.

Right - I should have said "every request that isn't...

```
 'api/*': lambdaBehavior,
  '_next/data/*': lambdaBehavior,
  // dynamic images go to lambda
  '_next/image*': imageBehavior,
  // known static routes
  // it would be nice to create routes for all the static files we know of
  // but we run into the limit of CacheBehaviors per distribution
  '_next/*': staticBehavior,
```
"

I wrote a VERY simple [locust](https://locust.io/) script to do some basic testing. **In the end, I'm not sure my testing methodology was enough to give me a ton of confidence that the test isn't flawed** - either that, or the test doesn't show a TON.

I did two things:

1. I used the Next.js app-playground project and generated 500 SSG pages, then I ran a load test against these pages (both with a fresh / empty cache as well as a pre-warmed cache) with both the S3 bucket as primary as well as the Lambda as primary
2. I ran the same tests, but against non SSG pages

Here is the locust test:

```
// locust.py
import time
import json
import random
from faker import Faker
fake = Faker()

from locust import HttpUser, task, between

class WebsiteUser(HttpUser):
  wait_time = between(1,5)

  def on_start(self):

    print ("done")
    return super().on_start()

  @task
  def list_page(self):
    url = "/ssg/" + str(random.randint(1, 500))    
    result = self.client.get(
            url=url,
            name="Get SSG Page",
            headers={
                "Accept": "text/html",
            },
            
        )
    print(result)
    ```
Ran with (for example) `locust -f locust/locust.py -u 15 -r 5 --run-time 300s --headless --host https://d2sstn6osjz1uu.cloudfront.net --html 300_second_GENERATED_pa
ges_lambda_first.html`    
 
Here are the test results:   
[PR-125-locust_tests_1.zip](https://github.com/jetbridge/cdk-nextjs/files/12234876/PR-125-locust_tests_1.zip)

Reply:
For SSG, it makes sense that S3 first would be faster. Now I'm second guessing the actual cause of my big latency by hitting S3 first for SSR.

Reply:
I suppose that's true. For these tests ("GENERATED" === non-SSG, these were rendered SSR though (I think?)), the difference between S3 first vs Lambda is "only" 23ms. If I'm honest though I don't have enough experience with this setup to know if that's meaningful, or noise.

![Screenshot 2023-08-01 at 9 06 45 PM](https://github.com/jetbridge/cdk-nextjs/assets/21270844/ef5796f5-7928-4253-a90c-28f8415b2c17)

Reply:
For SSG, that would be cached by CDN anyways right? So it would be better to prefer lambda first on SSR. Maybe we can do what Mischa suggested and look up the routes manifest to route accordingly for best performance.

Reply:
If it's only 20ms, then that's not worth doing... but I swear I was getting 300ms the last time I used it. Maybe Nextjs improved since then? Or maybe I did something really stupid before.

Reply:
@khuezy actually, re:

> For SSG, it makes sense that S3 first would be faster. 

The more I think about this, the more the numbers don't make sense to me. I don't think the SSG files exist in S3 anymore (at least not the first S3 bucket), right? So the first call to S3 I'd think would be a miss, because those SSG files only exist in the `cache` directory, and the S3 bucket setup as an origin in CloudFront are just the static asset files?

Reply:
yea you're right, the .html files are in the cache folder.

Reply:
> If it's only 20ms, then that's not worth doing... but I swear I was getting 300ms the last time I used it. Maybe Nextjs improved since then? Or maybe I did something really stupid before.

It would greatly depend on what region your S3 bucket is in, I imagine.

Reply:
Hmm good point, I'm in west coast and my app was deployed to us-east-1. I think I was running it on the edge too, so that DB round trip might be why I was getting bad latency.

Reply:
My assumption is that CloudFront is in us-east-1, so if your S3 bucket is there then the origin request from CF to S3 should be low latency enough to not greatly impact request time.
If we were to make a Lambda@Edge function that checked for a file in S3, that edge function would possibly have extra latency since it could be run from any AWS POP.

So my point is I would expect using S3 as the primary origin and falling back to lambda if we get a 403/404 to have reasonable performance. Which is why it was originally implemented this way. The 23ms latency to check the S3 origin first sounds pretty reasonable to me. 

Obviously all of these approaches have trade-offs. I think the maximum flexibility and robustness of trying S3 first is worth the extra latency for those requests, and it's always possible to provide some route prefixes that can be routed straight to lambda. Maybe that would be a handy configuration option: `directAppRoutes: string[]` or somesuch, which creates behaviors that skip the fallback origin and go straight to the lambda. Then as a user of this module I could plug in route prefixes that I know are not static files.

Reply:
Great work, @kevin-mitchell!
If it's only 10s of ms difference, I agree with @revmischa:
>  I think the maximum flexibility and robustness of trying S3 first is worth the extra latency for those requests

Reply:
Cool so we're all on the same consensus. If a user experiences longer latency due to edge roundtrip to a distant bucket, they can perhaps override the fallback.

Reply:
Sounds good! I went ahead and pushed a commit to swap the behavior here. This should resolve the issue with direct requests to static assets (e.g. `<img src="/blah.png" ... />` for `/public/blah.png`) failing.